// @stryker-mutator/util/test/unit/errors.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var errors_1 = require("../../src/errors");
var chai_1 = require("chai");
describe('errors', function () {
    describe('errorToString', function () {
        it('should return empty string if error is undefined', function () {
            chai_1.expect(errors_1.errorToString(undefined)).eq('');
        });
        it('should convert a nodejs Errno error to string', function () {
            var error = {
                code: 'foo',
                errno: 20,
                message: 'message',
                name: 'name',
                path: 'bar',
                stack: 'qux',
                syscall: 'baz'
            };
            chai_1.expect(errors_1.errorToString(error)).eq('name: foo (baz) qux');
        });
        it('should convert a regular error to string', function () {
            var error = new Error('expected error');
            chai_1.expect(errors_1.errorToString(error)).eq("Error: expected error\n" + (error.stack && error.stack.toString()));
        });
        it('should convert an error without a stack trace to string', function () {
            var error = new Error('expected error');
            delete error.stack;
            chai_1.expect(errors_1.errorToString(error)).eq('Error: expected error');
        });
    });
});
